include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+   *
+   * In case both approaches are coded, the deal.II style
+   * will be used.
+   *
+   * The second approach is more in style with the deal.II philosophy
+   * and it also allows command line customization, like for example,
+   * @code
+   * ./myApp -snes_type newtontr -ksp_type cg
+   * @endcode
+   * in case the user wants to change the default nonlinear solver to
+   * a trust region solver and iterate on the tangent system with CG,
+   * still using NonlinearSolver::solve_for_jacobian_system as a preconditioner.
+   *
+   * The first approach has instead the advantage that only the matrix assembly
+   * procedure has to be coded, thus allowing quicker implementations and
+   * faster turnaround for experimenting with linear solver preconditioning
+   * configurations via command line customizations, like for example,
+   * @code
+   * ./myApp -ksp_type cg -pc_type gamg
+   * @endcode
+   *
+   * In case the nonlinear equations are derived from energy minimization
+   * arguments, it may be beneficial to perform linesearch or test trust-region
+   * model reductions using the energy functional. In such cases, users can
+   * set an optional NonlinearSolver::energy callback.
+   */
+  template <typename VectorType  = PETScWrappers::VectorBase,
+            typename PMatrixType = PETScWrappers::MatrixBase,
+            typename AMatrixType = PMatrixType>
+  class NonlinearSolver
+  {
+  public:
+    /**
+     * Type that holds real-valued numbers.
+     *
+     * Used to represent norms.
+     */
+    using real_type = PetscReal;
+
+    /**
+     * Constructor.
+     */
+    NonlinearSolver(const NonlinearSolverData &data     = NonlinearSolverData(),
+                    const MPI_Comm &           mpi_comm = PETSC_COMM_WORLD);
+
+    /**
+     * Destructor.
+     */
+    ~NonlinearSolver();
+
+    /**
+     * Return the PETSc SNES object.
+     */
+    SNES
+    petsc_snes();
+
+    /**
+     * Return the underlying MPI communicator.
+     */
+    MPI_Comm
+    get_mpi_communicator() const;
+
+    /**
+     * Reset the solver, it does not change the customization.
+     */
+    void
+    reinit();
+
+    /**
+     * Reset solver.
+     * Change customization according to `data`.
+     */
+    void
+    reinit(const NonlinearSolverData &data);
+
+    /**
+     * Set the preconditioning matrix only.
+     *
+     * When used with NonlinearSolver::setup_jacobian and
+     * NonlinearSolver::solve_for_jacobian_system, PETSc will approximate the
+     * linear system matrix-vector product using an internal matrix-free
+     * representation.
+     *
+     * When used with NonlinearSolver::jacobian PETSc will use the same matrix
+     * for both preconditioning and matrix-vector products.
+     */
+    void
+    reinit_matrix(PMatrixType &P);
+
+    /**
+     * Set both the linear system matrix and the preconditioning matrix
+     * that PETSc will use.
+     */
+    void
+    reinit_matrices(AMatrixType &A, PMatrixType &P);
```

Comment:
```suggestion
    set_matrices(AMatrixType &A, PMatrixType &P);
```

Reply:
@bangerth If you are ok, I will also change the naming convention for the TS interface in this PR

Reply:
No worries. As I said, I first wanted to collect comments, apply all changes temporarily to this branch, and then I will open a separate PR.